### ISSUE TITLE
Getting Forbidden error when trying a path with a space in it

### ISSUE THREAD

**Issue by amergrgic** [pplu/aws-sdk-perl#359]
Testing on Ubuntu 18.04 with Paws 0.41

When I try to do a simple upload to S3 with a small script (see the script below), I get a Forbidden error, when the key doesn't have a space it does work. Don't know if it's related to #323, but it seems to be. This is a huge problem that I really need a workaround for. Anyone have an idea? I have tested with versions > 0.37 and all have this same issue. 

```perl
use Paws;

$ENV{'AWS_ACCESS_KEY'} = "xxx";
$ENV{'AWS_SECRET_KEY'} = "xxx";

my $s3 = Paws->service('S3', region => "eu-west-1");

$s3->PutObject(
        Bucket  => "bucketname",
        Key     => "text test.txt",
        Body    => "text.txt",
);
```

**Comment by byterock** [comment]
Hmm I have been working extensively on S3 paws  ofer the past few weeks.

If you have the opportunity.

Try this branch  of paws

https://github.com/byterock/aws-sdk-perl/tree/s3ObjectTagging

So you are saying that 

key     => "text test.txt",  

flails 

and

key     => "texttest.txt",

passes??


I will check that

**Comment by amergrgic** [comment]
@byterock yes, exactly. I have tried with the Paws CLI as well: 

```
paws S3 --region eu-west-1 PutObject Bucket: bucketname Key: "test text.txt" Body: text.txt
```

returns: 

```
Paws::S3 is not stable / supported / entirely developed at /usr/local/share/perl/5.26.1/Paws/S3.pm line 2.
The API retured an error: Forbidden(403)
```

**Comment by byterock** [comment]
Well first you need Git  installed ;)

https://github.com/byterock/aws-sdk-perl.git

git checkout s3ObjectTagging

Then follow the normal install dev steps

https://github.com/pplu/aws-sdk-perl#development-setup

this command will recompile just the S3 bits

carton exec builder-bin/gen_classes.pl --classes botocore/botocore/data/s3/2006-03-01/service-2.json

Alternatively if you really just want to hack about you can just drop these files

https://github.com/byterock/aws-sdk-perl/blob/s3ObjectTagging/lib/Paws/Net/RestXmlCaller.pm
https://github.com/byterock/aws-sdk-perl/blob/s3ObjectTagging/lib/Paws/Net/RestXMLResponse.pm

on the correct path of your Paws it should fix it

**Comment by amergrgic** [comment]
> Well first you need Git installed ;)
> 
> https://github.com/byterock/aws-sdk-perl.git
> 
> git checkout s3ObjectTagging
> 
> Then follow the normal install dev steps
> 
> https://github.com/pplu/aws-sdk-perl#development-setup
> 
> this command will recompile just the S3 bits
> 
> carton exec builder-bin/gen_classes.pl --classes botocore/botocore/data/s3/2006-03-01/service-2.json
> 
> Alternatively if you really just want to hack about you can just drop these files
> 
> https://github.com/byterock/aws-sdk-perl/blob/s3ObjectTagging/lib/Paws/Net/RestXmlCaller.pm
> https://github.com/byterock/aws-sdk-perl/blob/s3ObjectTagging/lib/Paws/Net/RestXMLResponse.pm
> 
> on the correct path of your Paws it should fix it

Yeah, I updated my comment, even your branch didn't fix the issue.

**Comment by byterock** [comment]
Opps did not see your update!    

Sorry for the above.  

Normally I hack into the two '.pm' files  and dump the 'response' and 'request'  objects to see what is really coming back.  I will be working on this shortly

**Comment by amergrgic** [comment]
> Opps did not see your update!
> 
> Sorry for the above.
> 
> Normally I hack into the two '.pm' files and dump the 'response' and 'request' objects to see what is really coming back. I will be working on this shortly

@byterock I tried btw replacing those two .pm files with the ones from your branch and I get a different error on my screen: 

```
The request signature we calculated does not match the signature you provided. Check your key and signing method.
```
That's with my testscript, Paws CLI gives me back: 

```
The API retured an error: The request signature we calculated does not match the signature you provided. Check your key and signing method.(SignatureDoesNotMatch)
```

**Comment by byterock** [comment]
I just ran that test and it worked fine  for me ;)

Which is good on my part but not on yours

Is that bucker of yorus public.  I can try and upload from my end?

**Comment by amergrgic** [comment]
> I just ran that test and it worked fine for me ;)
> 
> Which is good on my part but not on yours
> 
> Is that bucker of yorus public. I can try and upload from my end?

Bucket is not public unfortunately, did you test with a bucket that has a policy on it?

**Comment by byterock** [comment]
You are always on step ahead of me ;)

**Comment by byterock** [comment]
Yeah the code swap out was jsut a [Hail Mary Pass](https://en.wikipedia.org/wiki/Hail_Mary_pass) on my part :)

You most likly have to rebuild after a checkout of my branch.  

Could be a case of the policy being incorrect.

If you can go back to your .41 branch  and add in that dumper warning and compair to what we have above

**Comment by byterock** [comment]
Don't forget if you download my branch you still have to 'checkout' my branch

ie

git branch
  s3-todo-cleanup
* s3ObjectTagging

git checkout s3ObjectTagging

but you might already knwo that

**Comment by amergrgic** [comment]
@byterock good call, I think I forgot that the first time, but it still didn't help. The test you are doing, is that on Ubuntu 18.04 with Perl 5.26.1? I have a feeling it might be something Ubuntu related.

**Comment by amergrgic** [comment]
Update: Ubuntu 14.04 with Paws 0.41 works, but it encodes my filename to text%20text.txt

**Comment by amergrgic** [comment]
Ok, I have found the problem. The problem resides in using V4.pm in /usr/local/share/perl/5.26.1/Net/Amazon/Signature. The new version (0.19) doesn't seem to be working correctly with Paws > 0.37. When I moved V4.pm from a machine with Ubuntu 14.04 (which is version 0.16), my Paws works.

**Comment by byterock** [comment]
Glad I could help on that

Will see that bug might be my next blogging project S3 is coming to an end

**Comment by galenhuntington** [comment]
This looks the same as #323, where the issue is the encoded (rather than original) key being signed.  Have you tried my workaround there (altering `S3Signature.pm` to decode the `uri` field in-place before signing)?

**Comment by castaway** [comment]
It looks like we might need this PR applying to Signature::V4.. https://github.com/Grinnz/Net-Amazon-Signature-V4/pull/7

**Comment by Grinnz** [comment]
Net::Amazon::Signature::V4 0.20 has this change, please try updating. Thanks

**Comment by webtoolbox** [comment]
Upgrading Net::Amazon::Signature::V4 fixes the issue related to spaces in the object key, but it still fails in some cases. For example, with a file name like ~lkta100.gif.

**Comment by webtoolbox** [comment]
Upgrading Net::Amazon::Signature::V4 fixes the issue related to spaces in the object key, but it still fails in some cases. For example, with a file name like ~lkta100.gif.
